### ISSUE TITLE
Option to assign memory card readers to USB qube

### ISSUE THREAD

**Issue by andrewdavidwong** [QubesOS/qubes-issues#2055]
On my ThinkPad, the SD card reader is labeled "PCI Express Card Reader." Assigning it to my USB
qube results in any inserted SD card showing up in the USB qube (instead of dom0). However, I had to do this manually. It would be easier for most users to be presented an option for this to be done when the USB qube is first created.

Discussion: https://groups.google.com/d/topic/qubes-users/uX3GeQcqqTk/discussion

**Comment by marmarek** [comment]
Can you paste `lspci` description of this device?

**Comment by andrewdavidwong** [comment]
```
$ lspci -vv
02:00.0 Unassigned class [ff00]: Realtek Semiconductor Co., Ltd. RTS5227 PCI Express Card Reader (rev 01)
        Subsystem: Lenovo Device 5034
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0, Cache Line Size: 64 bytes
        Interrupt: pin A routed to IRQ 124
        Region 0: Memory at f1100000 (32-bit, non-prefetchable) [size=4K]
        Capabilities: <access denied>
        Kernel driver in use: pciback
        Kernel modules: rtsx_pci
```

**Comment by marmarek** [comment]
`Unassigned class` - doesn't look like invitation to make generic
solution for different card readers...

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by andrewdavidwong** [comment]
Seems like this should probably be closed in favor of #2454.

**Comment by andrewdavidwong** [comment]
My mistake. These two are actually separate (see https://github.com/QubesOS/qubes-issues/issues/2454#issuecomment-726056960).

**Comment by andrewdavidwong** [comment]
I know think it would be better to assign memory card readers to the USB qube by default, rather than relying on users to know about the option and opt in, so I've updated the issue title to reflect this.

**Comment by andrewdavidwong** [comment]
@marmarek, is this also relevant to security? Even though, as you pointed out in https://github.com/QubesOS/qubes-issues/issues/2454#issuecomment-726056960, memory cards are just storage devices, it still seems like a security risk to have them connect to dom0 for _some_ of the same reasons as USB storage devices.

**Comment by marmarek** [comment]
Generally yes, it is better to connect them to USB qube by default. But still the risk is much smaller than for USB, because here you have fixed device type instead of "anything" in case of USB. So, only filesystem parsing code + actual card reader driver code is exposed.
One practical issue is quite popular combined Realtek network + sd card reader device (why?!), which can't be separated and trigger https://www.qubes-os.org/doc/pci-troubleshooting/#libvirtlibvirterror-internal-error-unable-to-reset-pci-device--internal-error-active--devices-on-bus-with--not-doing-bus-reset . As you can see there, our current recommended solution is to simply disable one of them (preferably card reader in this case). If we assign SD card reader(s) to the USB qube by default, this issue will not only prevent sys-net from starting, but sys-usb too :/

**Comment by bettyvschmartz** [comment]
I have the same device as @andrewdavidwong but connecting through to any appVM from dom0 doesn't work. Is there some other PCI device you had to passthrough other than `00:02.0 Unassigned class [ff00]: Realtek Semiconductor Co., Ltd. RTS5227 PCI Express Card Reader (rev 01)` ? When I insert an SD card it shows up in dom0 in both `/dev` and with `lsblk` and I can get it to attach in the tray icon but in any appVM there is nothing in `/dev` or `lsblk` but `lspci` lists the device so it seems to be passed through. 

-- BVS

**Comment by andrewdavidwong** [comment]
> I have the same device as @andrewdavidwong but connecting through to any appVM from dom0 doesn't work. Is there some other PCI device you had to passthrough other than `00:02.0 Unassigned class [ff00]: Realtek Semiconductor Co., Ltd. RTS5227 PCI Express Card Reader (rev 01)` ? When I insert an SD card it shows up in dom0 in both `/dev` and with `lsblk` and I can get it to attach in the tray icon but in any appVM there is nothing in `/dev` or `lsblk` but `lspci` lists the device so it seems to be passed through.
> 
> -- BVS

No, I just assigned that one and both USB controller entries to my `sys-usb`, and I'm able to use SD cards just like USB storage drives.

**Comment by alimirjamali** [comment]
Catching up the discussion with @tlaurion on #9748

> `Unassigned class` - doesn't look like invitation to make generic
> solution for different card readers...

Then maybe it is better to do the opposite and only attach the tested and confirmed controllers (e.g. RTS5227). It appears that there are not many chips.

> One practical issue is quite popular combined Realtek network + sd card reader device (why?!), which can't be separated
... and we do not touch the trouble making ones. Do you remember the part number of this one?

I am looking at `qube-core-admin/qubes/device_protocol` and relevant places.

**Comment by tlaurion** [comment]
> > and whoever currently uses Card Reader in dom0 should stop doing so right away anyway and use the inconvenient ways of passing content to dom0 from sys-usb for damage control and prevention principles. My two cents.
> 
> We have to consider that some users format laptop internal storage with detached LUKs headers and keep `/boot` on external storage (USB sticks or SD CARD), mostly because they might want to hide their Qubes OS usage trace. And since SD CARDs have write-protect lock, some users might prefer to have `/boot` it on SD CARD.
> 
> p.s.: Attaching card readers to sys-usb by default should not break their existing Qubes OS installation for the mentioned users. They just won't be able to update their `/boot` easily if they desire to do so. Minor annoyance.

I'm not sure I follow why we would lower security for all when a small portion of people will use sdcard for /boot here and write protected (mechanical but also not guaranteed, see research papers on this, can be bypassed as well).

If this reasoning is good for those who use it, then I would recommend those users to swap out ssd drive if attached to an external bay, and if m2 drive then maybe reconsider this choice considering that card reader is exposed to dom0 which is a weird valid use case for those threat model (ro sdcard, but reader exposed to dom0 which makes dom0 parse partition table and fs, possibility leading to dom0 compromise with filesystem 0 day attack, which QubesOS is supposed to be designed to not ever expose dom0 to direct filesystem operations outside of passing lvm, raw disk images and never ever parsing things, outside of /boot it needs to trust.

**Comment by GrassfedMeatSticks** [comment]
I was also alarmed to see my SD Card pop up in Dom0 and have been following this Issue as well as a few others (https://github.com/QubesOS/qubes-issues/issues/3802, https://github.com/QubesOS/qubes-issues/issues/2454, https://github.com/QubesOS/qubes-issues/issues/1743, https://github.com/QubesOS/qubes-issues/issues/4235, https://github.com/QubesOS/qubes-issues/issues/9748, https://forum.qubes-os.org/t/how-to-hide-sd-controller-from-dom0-without-sys-usb/5801).

Two things I've tried that result in Qubes not detecting my SD Card at all:
1. Adding `rd.qubes.hide_pci=30:00.0` to the end of Dom0's `/etc/default/grub`'s `GRUB_CMDLINE_LINUX=` parameters and regenerating grub via `sudo grub2-mkconfig -o /boot/efi/EFI/qubes/grub.cfg` and rebooting.
2. Reverting the above change and adding my SD Card device to my **sys-usb** AppVM (both with no-strict-reset set to TRUE and FALSE) and rebooting.

Unlike other PCI devices mentioned in this thread, my SD Card appears as **30:00.0 SD Host controller: O2 Micro, Inc. SD/MMC Card Reader Controller (rev 01)** instead of the **Unassigned class** designation mentioned in this thread.

Am I missing something here, or is this what we're all seeing and discussing?
